File: src/libs/m3/vfs/GenericFile.cc
/*
 * Copyright (C) 2015, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/log/Lib.h>
#include <base/util/Time.h>

#include <m3/com/GateStream.h>
#include <m3/session/M3FS.h>
#include <m3/vfs/FileTable.h>
#include <m3/vfs/GenericFile.h>
#include <m3/Syscalls.h>

namespace m3 {

GenericFile::GenericFile(int flags, capsel_t caps, size_t id, epid_t mep, M3FS *sess_obj)
    : File(flags),
      _id(id),
      _sess_obj(sess_obj),
      _sess(caps + 0, sess_obj ? ObjCap::KEEP_CAP : 0),
      _sg(sess_obj ? &sess_obj->_gate : new SendGate(SendGate::bind(caps + 1))),
      _mg(MemGate::bind(ObjCap::INVALID)),
      _goff(),
      _off(),
      _pos(),
      _len(),
      _writing() {
    if(mep != EP_COUNT)
        _mg.ep(mep);
}

GenericFile::~GenericFile() {
    if(_writing)
        submit();

    if(flags() & FILE_NOSESS) {
        LLOG(FS, "GenFile[" << fd() << "," << _id << "]::close()");
        send_receive_vmsg(*_sg, M3FS::CLOSE_PRIV, _id);
        assert(_sess_obj);
        _sess_obj->free_ep(VPE::self().ep_to_sel(_mg.ep()));
    }
    else {
        if(_mg.ep() != MemGate::UNBOUND) {
            LLOG(FS, "GenFile[" << fd() << "," << _id << "]::revoke_ep(" << _mg.ep() << ")");
            capsel_t sel = VPE::self().ep_to_sel(_mg.ep());
            VPE::self().revoke(KIF::CapRngDesc(KIF::CapRngDesc::OBJ, sel), true);
            VPE::self().free_ep(_mg.ep());
        }
        delete _sg;
    }
}

Errors::Code GenericFile::stat(FileInfo &info) const {
    LLOG(FS, "GenFile[" << fd() << "," << _id << "]::stat()");

    GateIStream reply = send_receive_vmsg(*_sg, STAT, _id);
    reply >> Errors::last;
    if(Errors::last == Errors::NONE)
        reply >> info;
    return Errors::last;
}

ssize_t GenericFile::seek(size_t offset, int whence) {
    LLOG(FS, "GenFile[" << fd() << "," << _id << "]::seek(" << offset << ", " << whence << ")");

    // handle SEEK_CUR as SEEK_SET
    if(whence == M3FS_SEEK_CUR) {
        offset = _goff + _pos + offset;
        whence = M3FS_SEEK_SET;
    }

    // try to seek locally first
    if(whence == M3FS_SEEK_SET) {
        // no change?
        if(offset == _goff + _pos)
            return static_cast<ssize_t>(offset);

        // first submit the written data
        if(_writing && submit() != Errors::NONE)
            return -1;

        if(offset >= _goff && offset <= _goff + _len) {
            _pos = offset - _goff;
            return static_cast<ssize_t>(offset);
        }
    }
    else {
        // first submit the written data
        if(_writing && submit() != Errors::NONE)
            return -1;
    }

    // now seek on the server side
    size_t off;
    GateIStream reply = !have_sess() ? send_receive_vmsg(*_sg, SEEK, _id, offset, whence)
                                     : send_receive_vmsg(*_sg, SEEK, offset, whence);
    reply >> Errors::last;
    if(Errors::last != Errors::NONE)
        return -1;

    reply >> _goff >> off;
    _pos = _len = 0;
    return static_cast<ssize_t>(_goff + off);
}

ssize_t GenericFile::read(void *buffer, size_t count) {
    if(delegate_ep() != Errors::NONE)
        return -1;
    if(_writing && submit() != Errors::NONE)
        return -1;

    LLOG(FS, "GenFile[" << fd() << "," << _id << "]::read("
        << count << ", pos=" << (_goff + _pos) << ")");

    if(_pos == _len) {
        Time::start(0xbbbb);
        GateIStream reply = !have_sess() ? send_receive_vmsg(*_sg, READ, _id, static_cast<size_t>(0))
                                         : send_receive_vmsg(*_sg, READ, static_cast<size_t>(0));
        reply >> Errors::last;
        Time::stop(0xbbbb);
        if(Errors::last != Errors::NONE)
            return -1;

        _goff += _len;
        reply >> _off >> _len;
        _pos = 0;
    }

    size_t amount = Math::min(count, _len - _pos);
    if(amount > 0) {
        Time::start(0xaaaa);
        if(flags() & FILE_NODATA)
            CPU::compute(count / 2);
        else
            _mg.read(buffer, amount, _off + _pos);
        Time::stop(0xaaaa);
        _pos += amount;
    }
    return static_cast<ssize_t>(amount);
}

ssize_t GenericFile::write(const void *buffer, size_t count) {
    if(delegate_ep() != Errors::NONE)
        return -1;

    LLOG(FS, "GenFile[" << fd() << "," << _id << "]::write("
        << count << ", pos=" << (_goff + _pos) << ")");

    if(_pos == _len) {
        Time::start(0xbbbb);
        GateIStream reply = !have_sess() ? send_receive_vmsg(*_sg, WRITE, _id, static_cast<size_t>(0))
                                         : send_receive_vmsg(*_sg, WRITE, static_cast<size_t>(0));
        reply >> Errors::last;
        Time::stop(0xbbbb);
        if(Errors::last != Errors::NONE)
            return -1;

        _goff += _len;
        reply >> _off >> _len;
        _pos = 0;
    }

    size_t amount = Math::min(count, _len - _pos);
    if(amount > 0) {
        Time::start(0xaaaa);
        if(flags() & FILE_NODATA)
            CPU::compute(count / 4);
        else
            _mg.write(buffer, amount, _off + _pos);
        Time::stop(0xaaaa);
        _pos += amount;
    }
    _writing = true;
    return static_cast<ssize_t>(amount);
}

void GenericFile::evict() {
    assert(!(flags() & FILE_NOSESS));
    assert(_mg.ep() != MemGate::UNBOUND);
    LLOG(FS, "GenFile[" << fd() << "," << _id << "]::evict()");

    // submit read/written data
    submit();

    // revoke EP cap
    capsel_t ep_sel = VPE::self().ep_to_sel(_mg.ep());
    VPE::self().revoke(KIF::CapRngDesc(KIF::CapRngDesc::OBJ, ep_sel), true);
    _mg.ep(MemGate::UNBOUND);
}

Errors::Code GenericFile::submit() {
    if(_pos > 0) {
        LLOG(FS, "GenFile[" << fd() << "," << _id << "]::submit("
            << (_writing ? "write" : "read") << ", " << _pos << ")");

        GateIStream reply = !have_sess() ? send_receive_vmsg(*_sg, _writing ? WRITE : READ, _id, _pos)
                                         : send_receive_vmsg(*_sg, _writing ? WRITE : READ, _pos);
        reply >> Errors::last;
        if(Errors::last != Errors::NONE)
            return Errors::last;

        // if we append, the file was truncated
        size_t filesize;
        reply >> filesize;
        if(_goff + _len > filesize)
            _len = filesize - _goff;
        _goff += _pos;
        _pos = _len = 0;
    }
    _writing = false;
    return Errors::NONE;
}

Errors::Code GenericFile::delegate_ep() {
    if(_mg.ep() == MemGate::UNBOUND) {
        assert(!(flags() & FILE_NOSESS));
        epid_t ep = VPE::self().fds()->request_ep(this);
        LLOG(FS, "GenFile[" << fd() << "," << _id << "]::delegate_ep(" << ep << ")");
        _sess.delegate_obj(VPE::self().ep_to_sel(ep));
        if(Errors::last != Errors::NONE)
            return Errors::last;
        _mg.ep(ep);
    }
    return Errors::NONE;
}

}

File: src/libs/rustm3/src/vfs/genericfile.rs
/*
 * Copyright (C) 2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use cap::Selector;
use cell::RefCell;
use col::Vec;
use core::fmt;
use com::{MemGate, RecvGate, SendGate, SliceSource, VecSink};
use serialize::Sink;
use errors::Error;
use goff;
use io::{Read, Write};
use kif::{CapRngDesc, CapType, INVALID_SEL, Perm, syscalls};
use rc::Rc;
use session::{Pager, ClientSession};
use time;
use util;
use vfs;
use vpe::VPE;

int_enum! {
    pub struct Operation : u64 {
        const STAT      = 0;
        const SEEK      = 1;
        const NEXT_IN   = 2;
        const NEXT_OUT  = 3;
        const COMMIT    = 4;
    }
}

pub struct GenericFile {
    fd: vfs::Fd,
    sess: ClientSession,
    sgate: SendGate,
    mgate: MemGate,
    goff: usize,
    off: usize,
    pos: usize,
    len: usize,
    writing: bool,
}

impl GenericFile {
    pub fn new(sel: Selector) -> Self {
        GenericFile {
            fd: 0,
            sess: ClientSession::new_bind(sel),
            sgate: SendGate::new_bind(sel + 1),
            mgate: MemGate::new_bind(INVALID_SEL),
            goff: 0,
            off: 0,
            pos: 0,
            len: 0,
            writing: false,
        }
    }

    pub fn unserialize(s: &mut SliceSource) -> vfs::FileHandle {
        let _ : i32 = s.pop(); // flags
        Rc::new(RefCell::new(GenericFile::new(s.pop())))
    }

    fn submit(&mut self, force: bool) -> Result<(), Error> {
        if self.pos > 0 && (self.writing || force) {
            let mut reply = send_recv_res!(
                &self.sgate, RecvGate::def(),
                Operation::COMMIT, self.pos
            )?;
            // if we append, the file was truncated
            let filesize = reply.pop();
            if self.goff + self.len > filesize {
                self.len = filesize - self.goff;
            }
            self.goff += self.pos;
            self.pos = 0;
            self.len = 0;
            self.writing = false;
        }
        Ok(())
    }

    fn delegate_ep(&mut self) -> Result<(), Error> {
        if self.mgate.ep().is_none() {
            let ep = VPE::cur().files().request_ep(self.fd)?;
            self.sess.delegate_obj(VPE::cur().ep_sel(ep))?;
            self.mgate.set_ep(ep);
        }
        Ok(())
    }
}

impl vfs::File for GenericFile {
    fn fd(&self) -> vfs::Fd {
        self.fd
    }
    fn set_fd(&mut self, fd: vfs::Fd) {
        self.fd = fd;
    }

    fn evict(&mut self) {
        // submit read/written data
        self.submit(true).ok();

        // revoke EP cap
        let ep = self.mgate.ep().unwrap();
        let sel = VPE::cur().ep_sel(ep);
        VPE::cur().revoke(CapRngDesc::new(CapType::OBJECT, sel, 1), true).ok();
        self.mgate.unset_ep();
    }

    fn close(&mut self) {
        self.submit(false).ok();

        if let Some(ep) = self.mgate.ep() {
            let sel = VPE::cur().ep_sel(ep);
            VPE::cur().revoke(CapRngDesc::new(CapType::OBJECT, sel, 1), true).ok();
            VPE::cur().free_ep(ep);
        }

        // do that manually here, because the File object is not destructed if we've received this
        // file from our parent.
        VPE::cur().revoke(CapRngDesc::new(CapType::OBJECT, self.sess.sel(), 1), false).ok();
    }

    fn stat(&self) -> Result<vfs::FileInfo, Error> {
        let mut reply = send_recv_res!(
            &self.sgate, RecvGate::def(),
            Operation::STAT
        )?;
        Ok(reply.pop())
    }

    fn file_type(&self) -> u8 {
        b'F'
    }

    fn exchange_caps(&self, vpe: Selector,
                            _dels: &mut Vec<Selector>,
                            max_sel: &mut Selector) -> Result<(), Error> {
        let crd = CapRngDesc::new(CapType::OBJECT, self.sess.sel(), 2);
        let mut args = syscalls::ExchangeArgs::default();
        self.sess.obtain_for(vpe, crd, &mut args)?;
        *max_sel = util::max(*max_sel, self.sess.sel() + 2);
        Ok(())
    }

    fn serialize(&self, s: &mut VecSink) {
        s.push(&0); // flags
        s.push(&self.sess.sel());
        s.push(&0); // id
    }
}

impl vfs::Seek for GenericFile {
    fn seek(&mut self, mut off: usize, mut whence: vfs::SeekMode) -> Result<usize, Error> {
        self.submit(false)?;

        if whence == vfs::SeekMode::CUR {
            off = self.goff + self.pos + off;
            whence = vfs::SeekMode::SET;
        }

        if whence != vfs::SeekMode::END && self.pos < self.len {
            if off > self.goff && off < self.goff + self.len {
                self.pos = off - self.goff;
                return Ok(off)
            }
        }

        let mut reply = send_recv_res!(
            &self.sgate, RecvGate::def(),
            Operation::SEEK, off, whence
        )?;
        self.goff = reply.pop();
        let off: usize = reply.pop();
        self.pos = 0;
        self.len = 0;
        Ok(self.goff + off)
    }
}

impl Read for GenericFile {
    fn read(&mut self, buf: &mut [u8]) -> Result<usize, Error> {
        self.delegate_ep()?;
        self.submit(false)?;

        if self.pos == self.len {
            time::start(0xbbbb);
            let mut reply = send_recv_res!(
                &self.sgate, RecvGate::def(),
                Operation::NEXT_IN
            )?;
            time::stop(0xbbbb);
            self.goff += self.len;
            self.off = reply.pop();
            self.len = reply.pop();
            self.pos = 0;
        }

        let amount = util::min(buf.len(), self.len - self.pos);
        if amount > 0 {
            time::start(0xaaaa);
            self.mgate.read(&mut buf[0..amount], (self.off + self.pos) as goff)?;
            time::stop(0xaaaa);
            self.pos += amount;
        }
        self.writing = false;
        Ok(amount)
    }
}

impl Write for GenericFile {
    fn flush(&mut self) -> Result<(), Error> {
        self.submit(false)
    }

    fn write(&mut self, buf: &[u8]) -> Result<usize, Error> {
        self.delegate_ep()?;

        if self.pos == self.len {
            time::start(0xbbbb);
            let mut reply = send_recv_res!(
                &self.sgate, RecvGate::def(),
                Operation::NEXT_OUT
            )?;
            time::stop(0xbbbb);
            self.goff += self.len;
            self.off = reply.pop();
            self.len = reply.pop();
            self.pos = 0;
        }

        let amount = util::min(buf.len(), self.len - self.pos);
        if amount > 0 {
            time::start(0xaaaa);
            self.mgate.write(&buf[0..amount], (self.off + self.pos) as goff)?;
            time::stop(0xaaaa);
            self.pos += amount;
        }
        self.writing = true;
        Ok(amount)
    }
}

impl vfs::Map for GenericFile {
    fn map(&self, pager: &Pager, virt: goff, off: usize, len: usize, prot: Perm) -> Result<(), Error> {
        // TODO maybe check here whether self is a pipe and return an error?
        pager.map_ds(virt, len, off, prot, &self.sess).map(|_| ())
    }
}

impl fmt::Debug for GenericFile {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "GenFile[sess={}, goff={:#x}, off={:#x}, pos={:#x}, len={:#x}]",
            self.sess.sel(), self.goff, self.off, self.pos, self.len)
    }
}
